plugins/org.eclipse.n4js.ts.model/emf-gen/org/eclipse/n4js/ts/types/TModule.java

```diff
@@ -448,6 +449,37 @@
 	 */
 	EList<ComposedMemberCache> getComposedMemberCaches();
 
+	/**
+	 * Returns the value of the '<em><b>Temporary Types</b></em>' containment reference list.
+	 * The list contents are of type {@link org.eclipse.n4js.ts.types.Type}.
+	 * <!-- begin-user-doc -->
+	 * <!-- end-user-doc -->
+	 * <!-- begin-model-doc -->
+	 * *
+	 * As a rule, types must always be contained in a resource and resource set (as opposed to type references which
+	 * need not be contained). Therefore, when types are required to be created for a temporary purpose, they need to be
+	 * added somewhere, temporarily. This transient containment reference is used for that purpose.
+	 * <p>
+	 * IMPORTANT: code using this reference should
+	 * <ol>
+	 * <li>turn off notifications whenever adding/removing a type to this reference or when changing properties of an
+	 * already added type (to avoid unnecessary cache clear),
+	 * <li>remove the temporary type when it is no longer required,<br>
+	 * For convenience, the transpiler infrastructure will <b>clear this reference after transpilation</b> has
+	 * completed, so transpiler transformations do not have to handle removal of temporary types themselves.
+	 * <li>make sure no ordinary types ever refer to a temporary type contained here (references in the other direction
+	 * are ok, though),
+	 * <li>use the convenience methods for dealing with this reference provided in {@code N4JSResource}:
+	 * {@code addTemporaryType} and {@code clearTemporaryTypes}.
+	 * </ol>
+	 * <!-- end-model-doc -->
+	 * @return the value of the '<em>Temporary Types</em>' containment reference list.
+	 * @see org.eclipse.n4js.ts.types.TypesPackage#getTModule_TemporaryTypes()
+	 * @model containment="true" transient="true"
+	 * @generated
+	 */
+	EList<Type> getTemporaryTypes();
```

Comment:
As discussed, it is not 100% beautiful that we add a `ModuleNamespaceVirtualType` to this new `TemporaryType` in `TModule ` since in the "real" `TModule`, a `ModuleNamespaceVirtualType` is added to `internalTypes`. But I agree that it is OK for now.

Reply:
yeah, given the rule that types must be contained in a resource [set] we just have the choice between two evils, here. This seems like the lesser evil.